ex13_library : Online book store page.

Indigoinspn opened the pull request:
Created base landing page according to given design.
Book information (Image/Title/Author/Star rating) rendered through JS.

Indigoinspn committed:
ex13_library : Online book store page.
Created base landing page according to given design.
Book information (Image/Title/Author/Star rating) rendered through JS.

Indigoinspn committed:
Improved code for proccessing star rating.

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,150 @@
+
+var booksContainer = document.getElementById("books-container");
+
+function toPublishBooks(arr){    
```
зачем `to`? 🤔

Indigoinspn replied:
Чтобы показать, что функция выполняет действие) ( "публикует" книги на странице.

eugrdn replied:
обычно в названиях методов/функций не используют артикли, предлоги, частицы, не только в js, но и в других языках, по крайней мере я не встечал, возможно, чтобы не добавлять к и так непростой логике еще и грамматику 🙂 
лучше просто `publishBooks`, а еще лучше `renderBooks`

Indigoinspn replied:
хорошо=)

Indigoinspn replied:
Поправила название

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,150 @@
+
+function toPublishBooks(arr){    
+
+    for (var i=0; i< arr.length; i++) {
```
зачем все усложнять, если можно воспольоваться уже готовым методом `forEach`? 🙂

eugrdn replied:
и все, что ниже, для каждой книжки я бы вынес в отдельную функцию, возможно, в дальнейшем пригодится ее использование где-то еще

Indigoinspn replied:
Заменила на forEach метод и вынесла в отдельную функцию.

eugrdn reviewed (changes requested):
по верстке:
- нужно всю библиотеку обернуть в контейнер и вытянуть ее на 100% ширины
- прижать футер к низу страницы

можно так же в этой же ветке постараться сделать обработчики для рейтинга (сейчас + за половинчатые, но возможно будет сложно именно при обработке событий, тогда можешь оставить только полное заполнение)

Indigoinspn committed:
Extend content to the full width of the browser window.
Attach footer to the bottom of the browser window.

Indigoinspn commented:
Прижала футер к низу страницы и растянула контент на всю ширину.

Indigoinspn committed:
Working on user star rating. Added onclick event.

Indigoinspn committed:
Fix StarRating onclick event.
Useed delegation from main booksContainer. Defined target star array
through the NodeList. Recorded user rating to the dataBooks ("Json Array")

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+document.body.style.overflow = "hidden";
```
ой-ой
смысл такой: тебе объединить header, main и footer в один контейнер, указать у него высоту 100vh, ширину 100%, и дальше от этих размеров исходить отображая дочерние элементы, относительную высоту(можно в %), ширину (можно в %), отступы не задают

eugrdn replied:
все это делается средствами html и css

Indigoinspn replied:
Я вообще думала, что ты имеешь в виду, чтобы сайт занимал всю страницу и не было прокрутки). Когда я применяла 100vh страница занимала весь экран браузера, но при прокрутке под футером появлялось пустое место.  Еще хотела уточнить на макете сайта края закруглены, если сайт занимает 100% ширины, то в углях появляются белые просветы.

eugrdn replied:
> Я вообще думала, что ты имеешь в виду, чтобы сайт занимал всю страницу и не было прокрутки
да, я это и имел в виду, но ты сейчас просто обрезаешь то, что в нее не влезло, так нехорошо
можно без закруглений, там видимо скрин с macbook

Indigoinspn replied:
Могу я как общий контейнер использовать body?  У меня он в принципе и так 100vh.  Убрала из js первую строку (document.body.style.overflow = "hidden";)

eugrdn replied:
да, ок

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+  }
+};
+
+dataBooks.forEach(renderBooks);
```
forEach, т.е для каждого элемента dataBooks, те не renderBooks, a renderBook

Indigoinspn replied:
поправила

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+    "<div class=\"star-rating-container\"  data-book-id = \"" + item.id + "\" > </div></div>";
+
+  var starRate = item.rating;
+  var starRateContainer = booksContainer.getElementsByClassName("star-rating-container")[i];
```
а зачем тебе сначала добавлять книгу в dom, а потом тут же ее из него доставать, чтобы добавить рейтинг, если можно отрендерить сразу с ним??

Indigoinspn replied:
не совсем поняла, что ты имел ввиду и как лучше сделать =(. Как я могу отрендерить сразу с рейтингом? я же далее и определяю логику, чтобы отрендерить звезды.

eugrdn replied:
> booksContainer.innerHTML +=

добавила книгу в dom

> var starRateContainer =

получила рейтинг той же книги

ты же можешь

> booksContainer.innerHTML +=

перенести в самый конец функции и просто вставить рассчитанный рейтинг сюда

>  \+ item.id + "\" >___________ \</div>

Indigoinspn replied:
Попробовала так сделать, не получилось. Броузер сам добавляет закрывающий тег для контейнера звезд  (star-rating-container) и звезды оказываются за его пределами. Поэтому и приходится указать "var starRateContainer".

Indigoinspn replied:
Рассчитанный рейтинг сохранила в переменную и  потом добавила в booksContainer.innerHTML

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+  var starRate = item.rating;
+  var starRateContainer = booksContainer.getElementsByClassName("star-rating-container")[i];
+
+  var wholeFilledStars = Math.round(Math.floor(starRate));
```
Math.round(Math.floor( )) вынести в отдельную функцию

Indigoinspn replied:
Вынесла

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+/************ FUNCTIONS:  Fill or Clear star  *************/
+
+function fillStars(start, finish, arr) {  
+  if (start === 0 & finish === 0) { // if bookRating 0 and initial filledStarIndex = selectedStarIndex = 0 (=index of fist star half)
```
здесь наверное `&&`?
не понял где в этом условии проверяется, что `bookRating 0`

Indigoinspn replied:
Поправила

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+  stars = document.querySelectorAll(starSelector); // define target star array 
+
+  
+  findIndexOfSelectedStar(target,bookIndex);
```
эта функция не просто ищет индекс, она делает еще кучу всего, а возвращаемое значение как раз не используется

eugrdn replied:
нужно либо вынести из нее часть, которая изменяет звезды, либо переименовать

Indigoinspn replied:
Разбила на несколько, но появились другие проблемы ( скрыла от линтера, чтобы с тобой посоветоваться)

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+
+/***************************   User rating  *******************************/
+
+var bookId,
```
от этих глобавльных переменных лучше избавится, использовать только внутри функций

Indigoinspn replied:
Перенесла внутрь функции и столкнулась с проблемой передачи значению в другие функции.
Передаю в ф-ю displayUserStarRatingChoice  и в ней хочу переприсвоить значения, а линтер ругается. Всю голову сломала(, как быть?

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+  if (!target.classList.contains("star")) return; 
+  
+  bookIndex = target.parentNode.attributes[1].value - 1; // define selected book index   
+  getBookRating (bookIndex); // get rating from dataBooks array and save to filledStarIndex    
```
возвращаемое значение не используется

Indigoinspn replied:
Сохраняю теперь в  filledStarIndex = getBookRating(bookIndex); Еще проблема, ф-я findIndexOfSelectedStar(stars) должна что-то возвращать в конце, а мне нужно чтобы возвращала только нужное значение(. Что в общем return сделать?

Indigoinspn replied:
Поправила

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,267 @@
+
+function findIndexOfSelectedStar(target,bookIndex) {
+
+  for (var k = 0; k < stars.length; k++) {
```
лучше привести к массиву и использовать forEach

Indigoinspn replied:
Попробовала, не получилось. Мне нужно чтобы вернулся индекс выбранной звезды, а forEach возвращает undefined. Вернулась к обычному перебору(

eugrdn replied:
используй метод `.find`

Indigoinspn replied:
Использовала findIndex().

eugrdn reviewed (changes requested):
Комментарии в коде бывают нужны, но лучше когда их нет, как минимум закоменнтированного кода 🙂

Indigoinspn committed:
Redo structure of code. Have some issues.

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,251 @@
+
+function getBookRating(bookIndex) { 
+  var rating; 
+  if (dataBooks[bookIndex].rating === 0) rating = 0;     
```
давай здесь нормальным синтаксисом
`if () {

} else {

}`

Indigoinspn replied:
Сделала

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,251 @@
+    "<div class=\"book-author\"> by " + item.author.firstName + " " + item.author.lastName + "</div>" +
+    "<div class=\"star-rating-container\"  data-book-id = \"" + item.id + "\" > </div></div>";
+
+  var starRateContainer = booksContainer.getElementsByClassName("star-rating-container")[i]; 
```
здесь это делать не нужно, все, что у тебя ниже добавляется в starRateContainer.innerHTML, просто записывай в одну переменную, а потом вставь ее строкой между тегами div, где формируется booksContainer, в star-rating-container

Indigoinspn replied:
Сделала

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,251 @@
+  if (filledStarIndex === selectedStarIndex) {
+
+    if (filledStarIndex === 0) {
+      bookRatingValue = 0.5;    // eslint-disable-line
```
это значение нигде не используется и ниже тоже в условиях

Indigoinspn replied:
Поправила

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,251 @@
+
+function setBookRating(bookIndex, bookRatingValue) {
+  dataBooks[bookIndex].rating = bookRatingValue;
+  return bookRatingValue;
```
обычно set-методы не возвращают значений

Indigoinspn replied:
Поправила

eugrdn commented on src/ex13_library/main.js:
```diff
@@ -0,0 +1,251 @@
+    clearStars(selectedStarIndex + 1, filledStarIndex, stars);
+  }
+  setBookRating(bookIndex, bookRatingValue);
+  getBookRating(bookIndex);
```
зачем здесь вызывается?

Indigoinspn replied:
Убрала getBookRating(bookIndex), без него все работает)

Indigoinspn committed:
Main.js:
	Redo star rating (click + hover),
	Fixed the code according to the comments.
Style.css: add hover event to third block filters.
